Enable authenticating with remote app

mjrousos opened the pull request:
This PR enables remote authentication so that the ASP.NET Core app in an upgrade workflow can defer to the original ASP.NET app for user identity.

Fixes #15

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationAuthHandler.cs:
```diff
@@ -0,0 +1,107 @@
+    private readonly IRemoteAuthenticationService _authService;
+    private readonly IEnumerable<IRemoteAuthenticationResultProcessor> _resultProcessors;
+    private readonly ILogger<RemoteAuthenticationAuthHandler> _logger;
+    private RemoteAuthenticationResult? _remoteAuthenticationResult;
```
Are handlers a transient service? I thought they'd be singleton

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationAuthHandler.cs:
```diff
@@ -0,0 +1,107 @@
+    {
+        // If remote authentication is enabled for the endpoint, retrieve the remote authentication results and return
+        // an appropriate AuthenticateResult value.
+        if (Context.GetEndpoint()?.Metadata.GetMetadata<IRemoteAuthenticationMetadata>() is IRemoteAuthenticationMetadata metadata && metadata.Enabled)
```
```suggestion
        if (Context.GetEndpoint()?.Metadata.GetMetadata<IRemoteAuthenticationMetadata>() is { Enabled: true})
```

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationExtensions.cs:
```diff
@@ -0,0 +1,81 @@
+    /// </summary>
+    /// <param name="authenticationBuilder">The authentication builder to register the remote authentication handler with.</param>
+    /// <returns>The authentication builder updated with the remote authentication handler added.</returns>
+    public static AuthenticationBuilder AddRemote(this AuthenticationBuilder authenticationBuilder)
```
We've been referring to it as `RemoteApp` for session - does it make sense to stay consistent there?

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationHttpHandler.Framework.cs:
```diff
@@ -0,0 +1,62 @@
+        }
+        else
+        {
+            // If a user is logged in (using ASP.NET's usual authenticaiton mechanisms), return that claims principal.
```
```suggestion
            // If a user is logged in (i.e. using ASP.NET's usual authentication mechanisms), return that claims principal.
```

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationHttpHandler.Framework.cs:
```diff
@@ -0,0 +1,62 @@
+                context.Response.ContentType = "application/octet-stream";
+                claimsPrincipal.WriteTo(new BinaryWriter(context.Response.OutputStream));
+            }
+            else
```
should we return a different error code if there is a `context.User` but it is not a `ClaimsPrincipal`?

mjrousos replied:
That's a good question. I like that idea but I'm not sure what status code would make sense, though. Maybe 500, but that's pretty general. Maybe 409 or 422 but neither of those seems quite right.

mjrousos replied:
I like that approach I'll do that.

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationOptions.cs:
```diff
@@ -0,0 +1,73 @@
+        "Authorization"
+    };
+
+    public static readonly string[] DefaultResponseHeadersToForward = new[]
```
Do we need to make them public? Array values could be modified here - if we need it public, can you expose it as a `IReadOnlyCollection`?

mjrousos replied:
Good point. I suppose these don't need to be public, but it seems nice-to-have to have them available (maybe users will want to take existing default response headers to forward and add a couple to make their own list). I'll make these ImmutableArrays.

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationResultFactory.cs:
```diff
@@ -0,0 +1,65 @@
+        {
+            try
+            {
+                using var reader = new BinaryReader(await response.Content.ReadAsStreamAsync());
```
the stream should be disposed

mjrousos replied:
Fixed

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationResultFactory.cs:
```diff
@@ -0,0 +1,65 @@
+            }
+            catch (Exception exc)
+            {
+                _logger.LogWarning(exc, "Failed to deserialize remote authentication response as a claims principal despite success status code");
```
Do we want to silently ignore? Or do we want to log and throw?

mjrousos replied:
I think we don't need to throw here. Even if there's no ClaimsPrincipal, we can still report the status code and headers. Likely something's wrong on the server side, but I lean towards allowing the client to still work (just without remote auth) and providing log messages.

If it turns out that this comes up a lot and we need to make the failure more obvious, we could switch to throwing.

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationResultFactory.cs:
```diff
@@ -0,0 +1,65 @@
+        }
+        foreach (var headerName in headersToForward)
+        {
+            ret.ResponseHeaders.Add(headerName, response.Headers.GetValues(headerName));
```
This seems to iterate a few times over the headers via LINQ - can you just do a single pass?

mjrousos replied:
I simplified this (and fixed a bug)

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationService.cs:
```diff
@@ -0,0 +1,131 @@
+/// <summary>
+/// Service for authenticating a user by making an HTTP request on their behalf to a remote app.
+/// </summary>
+public class RemoteAuthenticationService : IRemoteAuthenticationService
```
services have for now been kept internal with only the interfaces (as needed) being exposed. Is there a reason to expose this service itself?

twsouthwick reviewed (commented):
Awesome! Can you add some documentation about what the flow here looks like? (kind of like the remote session stuff).

mjrousos commented:
Yep, for sure. I'll get some docs added today.

twsouthwick commented:
Question: Do we want this in a separate library? The session stuff was put in a separate package because it would force new dependencies brought in by System.Text.Json. Does this bring in new dependencies?

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/IRemoteAuthenticationService.cs:
```diff
@@ -0,0 +1,30 @@
+        /// Initializes the remote authentication service for the given scheme.
+        /// </summary>
+        /// <param name="scheme">The scheme whose configuration the remote authentication service should use for authenticating requests.</param>
+        void Initialize(AuthenticationScheme scheme);
```
That would work. I chose this approach because it mirrors what `AuthenticationHandler` (which depends on the service) does. Having the initialize here makes it straightforward to have `RemoteAuthenticationAuthHandler`'s initialize method call the service's initialize method.

An alternative would be to have the `RemoteAuthenticationAuthHandler` have a dependency on a service factory, instead, and create the service in its Initialize method.

I don't have a strong preference one way or the other. If you think the factory approach would be cleaner/more maintainable/whatever, let me know and I'll change it.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/NullableAttributes.cs:
```diff
@@ -0,0 +1,180 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Good point. It's already in the core web adapters project. I'll link to it there.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationModule.Framework.cs:
```diff
@@ -0,0 +1,39 @@
+        {
+            var context = ((HttpApplication)sender).Context;
+
+            if (string.Equals(context.Request.Path, _options.AuthenticationEndpointPath)
```
Sure. I'll move them here.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationResult.cs:
```diff
@@ -0,0 +1,43 @@
+    /// Gets a dictionary of auth-related headers that may need propagated back
+    /// to the caller if remote authentication fails.
+    /// </summary>
+    public IDictionary<string, IEnumerable<string>> ResponseHeaders { get; } = new Dictionary<string, IEnumerable<string>>();
```
Good call; switching.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationService.cs:
```diff
@@ -0,0 +1,131 @@
+
+        // Create a new HTTP request, but propagate along configured headers or cookies
+        // that may matter for authentication
+        var authRequest = new HttpRequestMessage();
```
Good catch. Fixed.

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RedirectUrlProcessor.cs:
```diff
@@ -0,0 +1,68 @@
+                        // Update the query strings to use the original request path as the ReturnUrl query string
+                        queryStrings[ReturnUrlQueryStringName] = context.Request.Path.Value;
+
+                        // .NET Core 3.1 doesn't have a QueryHelpers.AddQueryString that takes an argument of type Dictionary<string, StringValues>
```
Aren't we planning to remove 3.1 for reasons like this?

twsouthwick replied:
It's removed now - you can use all the .NET 6 goodness you want :)

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RedirectUrlProcessor.cs:
```diff
@@ -0,0 +1,68 @@
+    /// <returns></returns>
+    public Task ProcessAsync(RemoteAuthenticationResult result, HttpContext context)
+    {
+        if (result.ResponseHeaders.TryGetValue(LocationHeaderName, out var locationHeaders))
```
Set-Cookie headers can have a similar problem with their path and domain fields.

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationAuthHandler.cs:
```diff
@@ -0,0 +1,107 @@
+            if (authResult.User is not null)
+            {
+                var ticket = new AuthenticationTicket(authResult.User, Scheme.Name);
+                _logger.LogDebug("Authenticated user {UserName} based on remote authentication service", authResult.User.Identity?.Name);
```
PII, remove the username.

mjrousos replied:
Fixed

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationAuthHandler.cs:
```diff
@@ -0,0 +1,107 @@
+            Context.Response.StatusCode = authResult.StatusCode;
+            foreach (var header in authResult.ResponseHeaders)
+            {
+                Context.Response.Headers.Add(header);
```
IHeaderDictionary.Add is a problematic API we're discourage using. E.g. it throws if you have multiple headers with the same name.
```suggestion
                Context.Response.Headers.Append(header.Key, header.Value);
```

mjrousos replied:
Fixed

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationExtensions.cs:
```diff
@@ -0,0 +1,81 @@
+/// <summary>
+/// Helper methods for registering remote authentication services
+/// </summary>
+public static class RemoteAuthenticationExtensions
```
RemoteAuthentication* -> RemoteAppAuthentication* to avoid conflicts with existing types.

mjrousos replied:
Done

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationModule.Framework.cs:
```diff
@@ -0,0 +1,51 @@
+            if (string.Equals(context.Request.Path, _options.AuthenticationEndpointPath)
+                && context.Request.HttpMethod.Equals("GET", StringComparison.OrdinalIgnoreCase))
+            {
+                // TODO : Do we need to even check the API key for this? The handler only returns information
+                //        about the currently authenticated user.
```
Yes, otherwise you risk leaking PII

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationModule.Framework.cs:
```diff
@@ -0,0 +1,51 @@
+                {
+                    // Using 407 here (proxy authentication required) to differentiate from the scenario of
+                    // a valid API key but no authenticated user.
+                    context.Response.StatusCode = 407;
```
It looks like this status code would be blindly returned by the front end to the client. We don't want to send them 407s.
400 bad request or 500 internal server error.

mjrousos replied:
500 may be the best option since 400 implies something wrong with the user's request when really it's an issue with proxy setup. I'll use that.

mjrousos replied:
Thinking more about this, we don't want any of these responses to get back to the user. I'll leave this as 407 but update the client (`RemoteAppAuthenticationAuthHandler`) to understand 407 and log the issue but not propagate the status code back to the user.

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationOptions.cs:
```diff
@@ -0,0 +1,74 @@
+#endif
+{
+#if NETCOREAPP3_1_OR_GREATER
+    public static readonly IEnumerable<string> DefaultHeadersToForward = ImmutableArray.CreateRange(new[]
```
```suggestion
    public static readonly IEnumerable<string> DefaultRequestHeadersToForward = ImmutableArray.CreateRange(new[]
```

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationOptions.cs:
```diff
@@ -0,0 +1,74 @@
+    /// Gets or sets a list of request headers that should be forwarded to the remote app for authentication purposes. If no headers
+    /// are specified, all headers will be forwarded.
+    /// </summary>
+    public IList<string> HeadersToForward { get; set; } = new List<string>(DefaultHeadersToForward);
```
```suggestion
    public IList<string> RequestHeadersToForward { get; set; } = new List<string>(DefaultHeadersToForward);
```

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationService.cs:
```diff
@@ -0,0 +1,131 @@
+    // Add configured cookies to the request, or all cookies if none in particualr are specified
+    private static void AddCookies(IEnumerable<string> cookiesToForward, HttpRequest originalRequest, HttpRequestMessage authRequest)
+    {
+        IEnumerable<string> cookieNames = originalRequest.Cookies.Keys;
```
Why filter and re-build the cookies? It would be simpler and more efficient to forward the whole Cookie header using the header forwarding logic.

mjrousos replied:
I was thinking that some users may want to forward only specific cookies (auth-related ones). I'm fine pulling this out, though, and just treating the cookie header as another header to forward. We can always revisit if we hear feedback from customers that they need more control.

mjrousos replied:
I've removed this and added "Cookie" to the list of request headers forwarded by default. If we find users need the ability to forward some cookies but not others, we can revisit this.

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationHttpHandler.Framework.cs:
```diff
@@ -0,0 +1,51 @@
+        }
+        else
+        {
+            context.Response.StatusCode = 401;
```
How do you expect the module to set challenges like 301 with a location header? Or 401 with a www-authenticate header?

mjrousos replied:
At least for the OWIN scenarios I've tried so far, the OWIN middleware for cookie auth or bearer auth actually takes care of that. Those components just look for a 401 response (which I set here) and update headers, status code, etc. as needed. I still need to confirm that this will also "magically work" for forms auth scenarios but will look into that this week.

mjrousos replied:
I've confirmed that this works for forms auth, too, so at least for most common kinds of auth the 401 returned by the handler is translated into the appropriate status code and headers.

Tratcher reviewed (changes requested):
Which kinds of auth have you tried this with?

mjrousos commented:
@twsouthwick - that's a good point. This doesn't pull in any new dependencies, so we could include it in the SystemWebAdapters project without adding a new one. I suppose it might be nice for customers to not have too many packages they need to install.

@Tratcher - so far, I've tested it with OWIN cookie auth (ASP.NET Idenity) and it works for that scenario. I'll try out forms auth and JWT tokens to confirm it works for those scenarios, too.

davidfowl commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/IRemoteAuthenticationMetadata.cs:
```diff
@@ -0,0 +1,17 @@
+/// be enabled for an endpoint or not. Remote authentication enables useful migration
+/// scenarios but adds unnecessary performance overhead in cases where it isn't needed.
+/// </summary>
+public interface IRemoteAuthenticationMetadata
```
Why do we need this? Can't we treat it like any other auth handler?

davidfowl replied:
Can we remove this?

mjrousos replied:
Yea, I'll pull this out and we can just have options to either make this scheme the default scheme (which would be like enabling the metadata everywhere) or not (in which case, users will have to specify specific endpoints this scheme is used with).

mjrousos replied:
Ok, this is removed now. Instead, users specify whether to make remote auth the default or not when registering SystemWebAdapter services for it.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationOptions.cs:
```diff
@@ -0,0 +1,74 @@
+
+namespace Microsoft.AspNetCore.SystemWebAdapters.Authentication;
+
+public class RemoteAuthenticationOptions
```
I'll use RemoteAppAuthenticationOptions instead.

Tratcher commented on docs/remote-authentication/remote-authentication.md:
```diff
@@ -0,0 +1,81 @@
+
+In addition to configuring the remote app's URL and the shared secret ApiKey, the callback passed to `AddRemoteAppAuthentication` can also optionally specify some aspects of the remote authentication process's behavior:
+
+* `HeadersToForward`: This property contains headers that should be forwarded from a request when calling the authenticate API. By default, the only header forwarded is `Authorization`. If no headers are specified (including removing the default one), then all headers will be forwarded.
```
```suggestion
* `RequestHeadersToForward`: This property contains headers that should be forwarded from a request when calling the authenticate API. By default, the only header forwarded is `Authorization`. If no headers are specified (including removing the default one), then all headers will be forwarded.
```

Tratcher commented on docs/remote-authentication/remote-authentication.md:
```diff
@@ -0,0 +1,81 @@
+In addition to configuring the remote app's URL and the shared secret ApiKey, the callback passed to `AddRemoteAppAuthentication` can also optionally specify some aspects of the remote authentication process's behavior:
+
+* `HeadersToForward`: This property contains headers that should be forwarded from a request when calling the authenticate API. By default, the only header forwarded is `Authorization`. If no headers are specified (including removing the default one), then all headers will be forwarded.
+* `CookiesToForward`: This property contains cookies that should be forwarded from a request when calling the authenticate API. By default, all cookies are forwarded, but adding cookie names to this list will cause only those cookies to be used.
```
```suggestion
```

mjrousos commented:
@twsouthwick @Tratcher @davidfowl - Thanks for the reviews on this one. I think at this point I've addressed all the feedback so far. I've combined the auth code into the main System.Web adapters project and removed the remote auth metadata (depending, instead, on existing default scheme settings).

Can you please take another look and let me know what other feedback you have?

The only item still outstanding is that I need to test this with forms auth (I've tested it with cookies and bearer tokens). I'll work on that this week and get an additional sample added showing forms auth working.

twsouthwick commented on docs/remote-authentication/remote-authentication.md:
```diff
@@ -0,0 +1,80 @@
+
+* `RequestHeadersToForward`: This property contains headers that should be forwarded from a request when calling the authenticate API. By default, the only header forwarded is `Authorization`. If no headers are specified (including removing the default one), then all headers will be forwarded.
+* `ResponseHeadersToForward`: This property lists response headers that should be propagated back from the Authenticate request to the original call that prompted authentication in scenarios where identity is challenged. By default, this includes `Location`, `Set-Cookie`, and `WWW-Authenticate` headers.
+* `AuthenticationEndpointPath`: The endpoint on the ASP.NET app where authenticate requests should be made. This defaults to `/systemweb-adapters/authenticate` and must match the endpoint specified in ASP.NET authentication endpoint configuration.
```
```suggestion
* `AuthenticationEndpointPath`: The endpoint on the ASP.NET app where authenticate requests should be made. This defaults to `/systemweb-adapters/authenticate` and must match the endpoint specified in the  ASP.NET authentication endpoint configuration.
```

mjrousos replied:
Fixed

twsouthwick commented on samples/RemoteAuth/Bearer/RemoteBearer/App_Start/OpenIdConnectCachingSecurityTokenProvider.cs:
```diff
@@ -0,0 +1,93 @@
+using Microsoft.IdentityModel.Tokens;
+using Microsoft.Owin.Security.Jwt;
+
+// Coped from https://github.com/Azure-Samples/active-directory-b2c-dotnet-webapp-and-webapi
```
```suggestion
// Copied from https://github.com/Azure-Samples/active-directory-b2c-dotnet-webapp-and-webapi
```

mjrousos replied:
Fixed

twsouthwick commented on samples/RemoteAuth/Bearer/RemoteBearer/App_Start/OpenIdConnectCachingSecurityTokenProvider.cs:
```diff
@@ -0,0 +1,93 @@
+namespace RemoteOAuth
+{
+    // This class is necessary because the OAuthBearer Middleware does not leverage
+    // the OpenID Connect metadata endpoint exposed by the STS by default.
```
what's STS? should that be spelled out?

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ForwardedHost.Shared.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Microsoft.AspNetCore.SystemWebAdapters;
+
+internal struct ForwardedHost
```
```suggestion
internal readonly struct ForwardedHost
```

mjrousos replied:
Fixed.

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/IAuthenticationResultFactory.cs:
```diff
@@ -0,0 +1,23 @@
+/// Interface for creating authentication results from HTTP responses returned
+/// by a remote app in remote authentication scenarios.
+/// </summary>
+public interface IAuthenticationResultFactory
```
Does this need to be public? Or is it an implementation detail? Do we expect people to need to override this?

mjrousos replied:
I think it's unlikely users will override this, so I've made it internal.

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/IRemoteAppAuthenticationResultProcessor.cs:
```diff
@@ -0,0 +1,23 @@
+/// to specify heuristics to apply to remote authentication results to clean them up
+/// or modify their properties in an extensible way.
+/// </summary>
+public interface IRemoteAppAuthenticationResultProcessor
```
Do we expect people to need to implement this?

mjrousos replied:
This one, I think users may want to implement. If their ASP.NET app has any sort of custom authentication that would return different headers or status codes than expected, implementing this interface allows the user to tweak the response object accordingly.

twsouthwick commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationHttpHandler.Framework.cs:
```diff
@@ -0,0 +1,53 @@
+/// <summary>
+/// HTTP handler for serving requests to remote authenticaiton endpoint.
+/// </summary>
+internal sealed class RemoteAppAuthenticationHttpHandler : IHttpHandler
```
we could implement `IHttpHandler` and `IHttpModule` together for this to keep the code together. If not, could the API key check be moved to the module? That way it fails early

mjrousos replied:
Combining them would work in most cases, but I worry about edge cases where some module after this one interacts with auth. I think it would be better to leave them separate but make sure all API key stuff happens in the module.

twsouthwick reviewed (commented):
This looks great - had a few questions about visibility of some of the classes and some docs changes.

It would be great to see some tests for some of the processors to understand what it's expected to do. Not a blocker, though

mjrousos commented:
I've added a forms auth sample demonstrating that remote auth (more or less) works with forms auth, too. Like cookie auth, logging out from the ASP.NET Core app doesn't work (because forms auth logout does some magic that I haven't implemented on the ASP.NET Core side yet) but everything else works.

Tratcher commented:
> Like cookie auth, logging out from the ASP.NET Core app doesn't work (because forms auth logout does some magic that I haven't implemented on the ASP.NET Core side yet) but everything else works.

What kind of magic?

mjrousos commented:
The LoginStatus element of the forms app (`<asp:LoginStatus ID="LoginStatus1" runat="server" LogoutAction="Redirect" LogoutPageUrl="~/Logout.aspx" />`) generates a logout link that logs the user out via a post back (`__doPostBack('ctl00$LoginStatus1$ctl00','')`).

I expect the user could mimic this action from the ASP.NET Core app, but my guess is they'll need to make a request that looks like a post back and I haven't dug into what all that entails.

mjrousos commented:
Thanks, @Tratcher. Those are good questions. I agree that this isn't going to be the right approach for all auth scenarios (JWT can be done better pretty easily and we can potentially provide some code to help with cookies). I think this is a useful first step to getting auth working that can be used as a fallback for most scenarios (everything except Windows auth, I think) until we have a better solution or better guidance available.

As far as how to migrate off of it, that's a good point. I think we'll probably want a follow-up feature that allows users to authenticate in the *ASP.NET Core* app and forward the identity to the ASP.NET app (so, the reverse of what's happening here).  I've already prototyped that approach and it works. That can be a separate PR, though, and I think this one is the higher priority based on a principle of getting people started with .NET 6 as quickly and painlessly as possible.

mjrousos commented on samples/RemoteAuth/Bearer/RemoteBearer/packages.config:
```diff
@@ -0,0 +1,37 @@
+  <package id="Microsoft.IdentityModel.JsonWebTokens" version="6.18.0" targetFramework="net472" />
+  <package id="Microsoft.IdentityModel.Logging" version="6.18.0" targetFramework="net472" />
+  <package id="Microsoft.IdentityModel.Protocols" version="6.18.0" targetFramework="net472" />
+  <package id="Microsoft.IdentityModel.Protocols.OpenIdConnect" version="6.18.0" targetFramework="net472" />
+  <package id="Microsoft.IdentityModel.Protocols.WsFederation" version="5.3.0" targetFramework="net472" />
```
Huh, not sure. I'll update them all to 6.19.

mjrousos commented on samples/RemoteAuth/Bearer/RemoteBearer/App_Start/OpenIdConnectCachingSecurityTokenProvider.cs:
```diff
@@ -0,0 +1,93 @@
+            }
+        }
+
+        private void RetrieveMetadata()
```
Fixed

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/ResultProcessors/RedirectUrlProcessor.cs:
```diff
@@ -0,0 +1,116 @@
+{
+    private const string LocationHeaderName = "Location";
+    private const string ReturnUrlQueryStringName = "ReturnUrl";
+    private const string ForwardedHostHeaderName = "x-forwarded-host";
```
Not in `HeaderNames`, unfortunately, but I like the idea of using a shared const for this. I'll do that.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationService.cs:
```diff
@@ -0,0 +1,120 @@
+    private static void AddHeaders(IEnumerable<string> headersToForward, HttpRequest originalRequest, HttpRequestMessage authRequest)
+    {
+        // Add x-forwarded headers so that the authenticate API will
+        authRequest.Headers.Add("x-forwarded-host", originalRequest.Host.Value);
```
No, unfortunately.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationOptions.Shared.cs:
```diff
@@ -0,0 +1,69 @@
+namespace Microsoft.AspNetCore.SystemWebAdapters.Authentication;
+
+public class RemoteAppAuthenticationOptions
+#if NET6_0_OR_GREATER
```
That doesn't work here because we need to exclude .NET Standard, too.

mjrousos commented on docs/getting_started.md:
```diff
@@ -39,6 +39,10 @@ This step may require a number of projects to change depending on your solution
 
+## Enable shared authentication support
+
+While migrating endpoints from an ASP.NET app to an ASP.NET Core app, it is desirable to enable the user to log in to the application once (if authentication is supported) and for that identity to be available to both the ASP.NET and ASP.NET Core app so that the user's experience with the site works as it would if all endpoints were served from the same project. In order to enable this, the Syste ...
```
I think it's important we (briefly) discuss shared auth in the getting started and link to the page with more details. I'll see if I can simplify this.

mjrousos commented on docs/remote-authentication/remote-authentication.md:
```diff
@@ -0,0 +1,80 @@
+
+In addition to these options, `RemoteAuthenticationOptions` derives from `AuthenticationSchemeOptions`, so it can be used to optionally specify authentication schemes to forward to for different authentication actions.
+
+In addition to configuring services for the remote authentication handler, **endpoints must opt in to using remote authentication**. This can be done for an entire set of endpoints at once by calling `RequireRemoteAppAuthentication` while configuring the endpoints:
```
Good catch. I forgot to update this part of the docs when I made that change. I'll revise them.

mjrousos commented on docs/remote-authentication/remote-authentication.md:
```diff
@@ -0,0 +1,80 @@
+1. When requests are processed by the ASP.NET Core app, the `RemoteAuthenticationAuthHandler` will attempt to authenticate the user for the request.
+    1. The handler will first check whether the endpoint the request is being routed to has remote authentication metadata enabled (via a `[RemoteAuthentication]` or a call to `RequireRemoteAuthentication`). If remote authentication is not enabled, the handler will exit with a result of `NoResult`.
+    1. If remote authentication is enabled for the endpoint, the handler will make an HTTP request to the ASP.NET app's authenticate endpoint. It will copy configured headers and cookies from the current request onto this new one in order to forward auth-relevant data. As mentioned above, default behavior is to copy the `Authorize` header and all cookies. The API key header is also added for secu ...
+1. The ASP.NET app's `RemoteAuthenticationHttpHandler` will serve requests sent to the authenticate endpoint. As long as the API keys match, the handler will return either the current user's ClaimsPrincipal serialized into the response body (the use should have already been logged in thanks to the forwarded headers and cookies) or it will return 401. ASP.NET authentication middleware will interce ...
```
Removed

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/IRemoteAppAuthenticationService.cs:
```diff
@@ -0,0 +1,30 @@
+    /// <summary>
+    /// A service for authenticating an HTTP request with a remote service.
+    /// </summary>
+    public interface IRemoteAppAuthenticationService
```
No, good point. I'll make it internal.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationAuthHandler.cs:
```diff
@@ -0,0 +1,104 @@
+
+    private async Task<RemoteAppAuthenticationResult> GetRemoteAppAuthenticationResultAsync()
+    {
+        if (_remoteAppAuthResult is null)
```
We cache the result because we need the remote app authentication result both for `HandleAuthenticateAsync` and for `HandleChallengeAsync`, so this may be called twice.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationExtensions.cs:
```diff
@@ -0,0 +1,92 @@
+        authenticationBuilder.Services.AddSingleton<IAuthenticationResultFactory, RemoteAppAuthenticationResultFactory>();
+        authenticationBuilder.Services.AddHttpClient<IRemoteAppAuthenticationService, RemoteAppAuthenticationService>()
+            // Disable cookies in the HTTP client because the service will manage the cookie header directly
+            .ConfigurePrimaryHttpMessageHandler(() => new HttpClientHandler { UseCookies = false, AllowAutoRedirect = false });
```
For configuring an HTTP client with cookies and redirect disabled? I don't know that's necessary; this is the only place we set up an HttpClient exactly like this.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationOptions.Shared.cs:
```diff
@@ -0,0 +1,69 @@
+#endif
+    public string AuthenticationEndpointPath { get; set; } = "/systemweb-adapters/authenticate";
+
+#if NET6_0_OR_GREATER
```
Good idea. Fixed.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationResultFactory.cs:
```diff
@@ -0,0 +1,59 @@
+        // Copy expected response headers
+        foreach (var responseHeader in response.Headers)
+        {
+            if (!options.ResponseHeadersToForward.Any() || options.ResponseHeadersToForward.Contains(responseHeader.Key))
```
Sure. Fixed.

mjrousos commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationService.cs:
```diff
@@ -0,0 +1,120 @@
+    private readonly IAuthenticationResultFactory _resultFactory;
+    private readonly ILogger<RemoteAppAuthenticationService> _logger;
+    private readonly IOptionsMonitor<RemoteAppAuthenticationOptions> _optionsMonitor;
+    private RemoteAppAuthenticationOptions? _options;
```
Good point. I've made it an IOptionsSnapshot.

davidfowl commented on samples/RemoteAuth/Bearer/RemoteBearerCore/Program.cs:
```diff
@@ -0,0 +1,31 @@
+builder.Services.AddControllers();
+
+builder.Services.AddSystemWebAdapters()
+    .AddRemoteAppAuthentication(true, options =>
```
What is the `true` parameter?

mjrousos replied:
That's specifying whether remote app auth should be the default scheme or not. I updated the docs about that, but I'll add a comment here, as well.

Tratcher commented on src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationService.cs:
```diff
@@ -0,0 +1,120 @@
+    // Add configured headers to the request, or all headers if none in particular are specified
+    private static void AddHeaders(IEnumerable<string> headersToForward, HttpRequest originalRequest, HttpRequestMessage authRequest)
+    {
+        // Add x-forwarded headers so that the authenticate API will
```
will what?

mjrousos commented:
> This is the main thing we need guidance for, we can't strand people in this hybrid state.

Agreed. Figuring out how to migrate off of forms auth is orthogonal to this PR, though, since this is just about bridging the gap so that the apps can share auth while endpoints migrate over gradually. Having a path forward for folks who used forms auth will be important, though, whether a user is taking advantage of these tools or just migrating all at once on their own.

CZEMacLeod commented:
> > As far as how to migrate off of it, that's a good point. I think we'll probably want a follow-up feature that allows users to authenticate in the _ASP.NET Core_ app and forward the identity to the ASP.NET app (so, the reverse of what's happening here).
> 
> I don't know how many people will want that given they'll want to limit changes/investments in the old app. Have any customers we've interviewed expressed interest in that?
> 
> > I realize that doesn't make migrating off of forms auth any easier - that's a separate problem
> 
> This is the main thing we need guidance for, we can't strand people in this hybrid state.

I can state for a fact, that I would be interested in moving the auth to core, and forwarding to my asp.net app.
There are a number of improvements in core auth, and leaving it to last (e.g. move everything but auth to core, so that is all the asp.net app is doing then moving that) seems a bad way of progressing to me.

You already have to make changes to your old app (to handle session state etc.), so I don't see anything wrong with adding something else to provide auth sharing. I think there is already some stuff to handle this in `Microsoft.Owin.Security.Interop` with the `DataProtectorShim`, `AspNetTicketDataFormat`, and `ChunkingCookieManager` - although that probably only applies if you are using OWIN on both sides.

The only other thing I would say, is that being able to migrate my customized `Microsoft.AspNet.Identity.Core` / `Microsoft.AspNet.Identity.Owin` implementation without breaking my database format etc. would be a major help - but probably outside the scope of this project.
